lib/rgeo/cartesian/feature_methods.rb

```diff
@@ -50,31 +50,161 @@ def segments
-        return true if len == 1
-        return segments[0].s != segments[1].e if len == 2
-        segments.each_with_index do |seg, index|
-          nindex = index + 1
-          nindex = nil if nindex == len
-          return false if nindex && seg.contains_point?(segments[nindex].e)
-          pindex = index - 1
-          pindex = nil if pindex < 0
-          return false if pindex && seg.contains_point?(segments[pindex].s)
-          next unless nindex
-          oindex = nindex + 1
-          while oindex < len
-            oseg = segments[oindex]
-            return false if !(index == 0 && oindex == len - 1 && seg.s == oseg.e) && seg.intersects_segment?(oseg)
-            oindex += 1
-          end
-        end
-        true
+        # TODO: should we replace with graph.incident_edges.length == segments.length
+        # since graph isn't used elsewhere yet it might be more overhead.
+        li = SweeplineIntersector.new(segments)
+        li.proper_intersections.size.zero?
       end
 
       def length
         segments.inject(0.0) { |sum, seg| sum + seg.length }
       end
+
+      def crosses?(rhs)
+        case rhs
+        when Feature::LineString
+          crosses_line_string?(rhs)
+        else
+          super
+        end
+      end
+
+      private
+
+      # Determines if a cross occurs with another linestring.
+      # Process is to get the number of proper intersections in each geom
+      # then overlay and get the number of proper intersections from that.
+      # If the overlaid number is higher than the sum of individual self-ints
+      # then there is an intersection. Finally, we need to check the intersection
+      # to see that it is not a boundary point of either segment.
+      #
+      # @param rhs [Feature::LineString]
+      #
+      # @return [Boolean]
+      def crosses_line_string?(rhs)
+        self_ints = SweeplineIntersector.new(segments).proper_intersections
+        self_ints += SweeplineIntersector.new(rhs.segments).proper_intersections
+        overlay_ints = SweeplineIntersector.new(segments + rhs.segments).proper_intersections
+
+        (overlay_ints - self_ints).each do |int|
+          s1s = int.s1.s
+          s1e = int.s1.e
+          s2s = int.s2.s
+          s2e = int.s2.e
+          return true unless [s1s, s1e, s2s, s2e].include?(int.point)
+        end
+
+        false
+      end
+    end
+
+    module PolygonMethods
+      def graph
+        @graph ||= GeometryGraph.new(self)
+      end
+
+      private
+
+      # Similar implementation to the default version from the ValidOp
+      # module, but it overrides a few methods in favor of a graph for
+      # easier checks. Eventually most of these checks can be overriden
+      # and rely on the graph.
+      #
+      # TODO: will update these with more checks once we decide
+      # if this structure makes sense or if there's more metaprogramming
+      # we could do.
+      def check_valid_polygon
```

Comment:
I don't understand why you copy this method: shouldn't the _leaf_ methods only be overriden?

I mean that since `check_consistent_area` and `check_connected_interiors` are overriden, you don't need to override that bit.

Reply:
The reason I did that is that in the originally implemented method in `ImplHelper` the leaf methods accept the geometry as an input, while these can just use self. But that was causing an issue where I'd be passing an argument into the new leaf methods that don't expect one.

I can change the leaf methods to something like this `def check_connected_interiors(_)` so that they won't break upon receiving an argument. What are your thoughts @BuonOmo?

Reply:
I think I'm doing something weird with the OO aspect of this. Since I'm including the `ValidOp` module every geometry instance will get these methods, but the leaf methods are designed to be more of a function than a method in the sense that they need the object they're checking passed in as a parameter and aren't checking the caller.

For example, a polygon calling `check_connected_interior` is essentially doing this

```rb
polygon.check_connected_interior(polygon)
```

where I feel like it should just be

```rb
polygon.check_connected_interior
```

Maybe I need to break this down into more submodules like `ValidOp::PolygonChecks` and include those into the relevant classes?

Reply:
I broke it up into `ValidOp` and `ValidOpHelpers` where the helpers contain the functions and `ValidOp` has just the `valid?`, `invalid_reason`, and `check_valid/check_valid_*` methods.

I also show how the helpers can be overridden for custom implementations of the leaf functions (`lib/cartesian/valid_op.rb`). The new helpers module has to "inherit" the default implementation and re-implement any methods it needs then the new valid op module defines a method which specifies the helper module it will make calls to. It is a bit of a weird pattern and I can clarify more if need be, but it seems to provide flexibility with minimal code duplication.